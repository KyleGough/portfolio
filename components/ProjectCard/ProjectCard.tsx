import { FadeIn } from '@components/FadeIn';
import { getFormattedDate, getShortDate } from '@utilities/date';
import { Date } from '@utilities/types';
import { clsx } from 'clsx';
import Image, { StaticImageData } from 'next/image';
import Link from 'next/link';
import React from 'react';

interface ProjectCardProps {
  className?: string;
  image: StaticImageData;
  alt: string;
  date: Date;
  link: string;
  logo: string;
  logoAlt: string;
}

export const ProjectCard: React.FC<ProjectCardProps> = ({
  className,
  image,
  alt,
  date,
  link,
  logo,
  logoAlt,
}) => {
  return (
    <Link href={link}>
      <div
        className={clsx(
          'group overflow-hidden',
          'rounded-2xl border-2 border-white',
          'shadow drop-shadow-lg cursor-pointer'
        )}
      >
        <FadeIn className={className}>
          <div className="w-full">
            <Image
              src={image.src}
              alt={alt}
              width={image.width}
              height={image.height}
              placeholder="blur"
              blurDataURL={image.blurDataURL}
            />
          </div>
          <div className="text-white p-4 font-bold group-hover:brightness-125 group-focus:brightness-125 bg-nav-light">
            <div className="flex justify-between items-center">
              <p className="text-sm opacity-80">
                <time dateTime={getShortDate(date)}>
                  {getFormattedDate(date)}
                </time>
              </p>
              <Image src={logo} alt={logoAlt} width={30} height={30} />
            </div>
          </div>
        </FadeIn>
      </div>
    </Link>
  );
};
