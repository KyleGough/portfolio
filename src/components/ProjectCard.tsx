import { Link } from 'react-router-dom';
import ProjectChip from './ProjectChip';
import useObserveElement from '../hooks/useObserveElement';

export default function ProjectCard(props: { className?: string, src: string, alt: string, title: string, tagline: string, date: string, link: string, chipText: string }) {
    const [elementRef, isVisible] = useObserveElement<HTMLAnchorElement>({
        root: null,
        rootMargin: '0px',
        threshold: 0.5
    });

    return (
        <Link ref={elementRef} to={props.link} className={`${props.className ? props.className : ''} ${isVisible ? 'opacity-100' : 'opacity-0'} duration-1000 transition-opacity ease-in-out rounded-2xl overflow-hidden border-2 border-white shadow drop-shadow-lg group`}>
            <img className='w-full' src={props.src} alt={props.alt} />
            <div className='text-white p-4 font-bold group-hover:brightness-125 group-focus:brightness-125 bg-nav-light'>
                <h3 className='text-xl font-medium mb-4 h-[3.5rem] line-clamp-2'>{props.tagline}</h3>
                <div className='flex justify-between items-center'>
                    <p className='text-sm opacity-80'>{props.date}</p>
                    <ProjectChip name={props.chipText} />
                </div>
            </div>
        </Link>
    );
}