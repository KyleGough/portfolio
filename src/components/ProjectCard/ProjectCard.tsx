import { clsx } from 'clsx';
import React from 'react';
import { Link } from 'react-router-dom';

import { useObserveElement } from '../../hooks/useObserveElement';
import { getLongDate } from '../../utilities/date';
import { ProjectChip } from '../ProjectChip';

interface ProjectCardProps {
  className?: string;
  src: string;
  alt: string;
  tagline: string;
  date: string;
  link: string;
  chipText: string;
}

export const ProjectCard: React.FC<ProjectCardProps> = ({
  className,
  src,
  alt,
  tagline,
  date,
  link,
  chipText,
}) => {
  const [elementRef, isVisible] = useObserveElement<HTMLAnchorElement>({
    threshold: 0.5,
  });

  return (
    <Link
      ref={elementRef}
      to={link}
      className={clsx(
        className,
        { 'opacity-100': isVisible },
        { 'opacity-0': !isVisible },
        'group overflow-hidden',
        'duration-1000 transition-opacity ease-in-out',
        'rounded-2xl border-2 border-white',
        'shadow drop-shadow-lg'
      )}
    >
      <img className="w-full" src={src} alt={alt} />
      <div className="text-white p-4 font-bold group-hover:brightness-125 group-focus:brightness-125 bg-nav-light">
        <h3 className="text-xl font-medium mb-4 h-[3.5rem] line-clamp-2">
          {tagline}
        </h3>
        <div className="flex justify-between items-center">
          <p className="text-sm opacity-80">
            <time dateTime={date}>{getLongDate(date)}</time>
          </p>
          <ProjectChip name={chipText} />
        </div>
      </div>
    </Link>
  );
};
