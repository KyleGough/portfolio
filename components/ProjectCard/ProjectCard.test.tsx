import mockIntersectionObserver from '@mocks/mockIntersectionObserver';
import { render, screen } from '@testing-library/react';
import React from 'react';

import { ProjectCard } from './ProjectCard';

describe('ProjectCard component', () => {
  beforeEach(mockIntersectionObserver);

  it('renders', async () => {
    const { container } = render(
      <ProjectCard
        image={{
          src: '/#image',
          width: 100,
          height: 100,
          blurDataURL: '/#blur',
        }}
        alt="Alt Text"
        date={{ month: 4, year: 2022 }}
        link="/#project"
        logo="/#logo"
        logoAlt="Logo"
      />
    );

    expect(screen.getByText('April 2022')).toBeVisible();
    expect(container).toMatchSnapshot();
  });
});
