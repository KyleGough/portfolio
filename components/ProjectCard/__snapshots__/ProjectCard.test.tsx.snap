// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard component renders 1`] = `
<div>
  <a
    href="/#project"
  >
    <div
      class="group overflow-hidden rounded-2xl border-2 border-white shadow drop-shadow-lg cursor-pointer"
    >
      <div
        class="opacity-0 transition-opacity duration-1000"
      >
        <div
          class="w-full"
        >
          <img
            alt="Alt Text"
            data-nimg="1"
            decoding="async"
            height="100"
            loading="lazy"
            src="/_next/image?url=%2F%23image&w=256&q=75"
            srcset="/_next/image?url=%2F%23image&w=128&q=75 1x, /_next/image?url=%2F%23image&w=256&q=75 2x"
            style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
            width="100"
          />
        </div>
        <div
          class="text-white p-4 font-bold group-hover:brightness-125 group-focus:brightness-125 bg-nav-light"
        >
          <div
            class="flex justify-between items-center"
          >
            <p
              class="text-sm opacity-80"
            >
              <time
                datetime="2022-04"
              >
                April 2022
              </time>
            </p>
            <div
              class="brightness-200 text-xs px-4 py-0.5 h-8 leading-chip rounded-2xl font-extrabold text-chip bg-chip-light"
            >
              Haskell
            </div>
          </div>
        </div>
      </div>
    </div>
  </a>
</div>
`;
